src/storage-controller/src/lib.rs

```diff
@@ -2999,35 +2997,18 @@ where
                         // from the finalization register, unfortunately... but the next run
                         // of `finalize_shards` should notice the upper has advanced and tidy
                         // up.
+                        let persist_client = persist_client.clone();
                         mz_ore::task::spawn(|| format!("finalize_shard({shard_id})"), async move {
-                            let result =
-                                tokio::time::timeout(
-                                    Duration::from_secs(15 * 60),
-
-                                    write_handle.append(
-                                        Vec::<(
-                                            (mz_storage_types::sources::SourceData, ()),
-                                            T,
-                                            Diff,
-                                        )>::new(),
-                                        write_handle.upper().clone(),
-                                        Antichain::new(),
-                                    )
-                                ).await;
-
-                            // Rather than error, just leave this shard as one to finalize later.
-                            match result {
-                                Err(_) => {
-                                    warn!("timed out while trying to finalize shard {shard_id}");
-                                }
-                                Ok(Err(usage)) => {
-                                    error!("invalid usage while finalizing shard {shard_id}: {usage:?}")
-                                }
-                                Ok(Ok(Err(mismatch))) => {
-                                    warn!("unable to advance the upper of shard {shard_id} to the empty antichain: {mismatch:?}")
-                                }
-                                Ok(Ok(Ok(()))) => {}
-                            };
+                            let result = persist_client
+                                .finalize_shard::<SourceData, (), T, Diff>(
+                                    shard_id,
+                                    Diagnostics::from_purpose("finalizing shards"),
+                                )
+                                .await;
+
+                            if let Err(e) = result {
```

Comment:
I think the "Rather than error, just leave this shard as one to finalize later." comment that got dropped here is useful and would prefer that it stick around